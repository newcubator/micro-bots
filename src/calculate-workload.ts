import {MocoActivityResponse, MocoEmploymentsResponse, MocoSchedulesResponse} from "./types/moco-types";
import dayjs from "dayjs";
import {DayObjectType, WorkloadType} from "./types/workload-types";

export function calculateWorkload(duration: number) {
  return (input: [MocoSchedulesResponse, MocoEmploymentsResponse, MocoActivityResponse]): WorkloadType => {
    const [schedules, employments, activities] = input;

    if (employments.data.length == 0) return;

    const activitiesMap = activities.data.map(activity => {
      return {id: activity.id, date: activity.date, hours: activity.hours}
    })

    const pattern = employments.data[0].pattern;

    let expectedSum = 0;
    let workedSum = 0;
    let holidays = 0;
    let days: DayObjectType[] = []

    for (let step = 1; step <= duration; step++) {
      const day = dayjs().subtract(step, 'day');
      const dayString = day.format('YYYY-MM-DD');
      let dayObject: DayObjectType = {
        day: dayString,
        expectedHours: 0,
        worked: 0
      }

      // don't expect people to work on weekends
      if (day.get('day') == 0 || day.get('day') == 6) {
        dayObject.weekend = true
      }

      // ignore scheduled absence days
      const scheduled = schedules.data.find(schedule => schedule.date == dayString);
      if (scheduled) {
        if ((scheduled.assignment.name == "Urlaub" && !dayObject.weekend) || (scheduled.assignment.name == "Feiertag" && !dayObject.weekend)) {
          dayObject.holiday = true
          holidays++
        }
      }

      if (!dayObject.weekend && !dayObject.holiday) {
        let expected = pattern.am[day.get('day') - 1] + pattern.pm[day.get('day') - 1]
        dayObject.expectedHours = expected;
        expectedSum += expected
      }

      let workedActivities = activitiesMap.filter(activity => activity.date == dayString)
      if (workedActivities) {
        let workedHours = 0
        workedActivities.forEach((activity) => {
          workedHours += activity.hours
        })
        dayObject.worked = workedHours
        workedSum += workedHours
      }
      days.push(dayObject)
    }

    return {
      user: employments.data[0].user,
      expectedHours: expectedSum,
      holidays: holidays,
      workedHours: workedSum,
      days: days
    };
  }
}
