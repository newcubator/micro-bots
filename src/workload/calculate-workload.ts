import dayjs, { Dayjs } from 'dayjs';
import { MocoActivity, MocoEmployment, MocoSchedule } from '../moco/types/moco-types';
import { DayObjectType, WorkloadType } from '../moco/types/workload-types';

const DATE_FORMAT = 'YYYY-MM-DD';

const isWeekday = (day: Dayjs) => (day.get('day') > 0 && day.get('day') < 6);

const createDayRange = (from: Dayjs, to: Dayjs) => new Array(to.diff(from, 'day') + 1)
    .fill(from)
    .map((date, idx) => date.add(idx, 'day'));

const extractAbsenceHours = (schedules: MocoSchedule[], assignmentType?: string, day?: Dayjs) => {
    return schedules
        .filter((schedule: MocoSchedule) => !day || schedule.date === day.format(DATE_FORMAT))
        .filter((schedule: MocoSchedule) => !assignmentType || schedule.assignment.name === assignmentType)
        .reduce((sum, schedule: MocoSchedule) => sum + ((+schedule.am + +schedule.pm) * 4), 0);
};

// This function assumes that the pattern array always contains 5 entries, even if the employment ends in the middle of the week
const extractExpectedHours = (employment: MocoEmployment, day: Dayjs) => employment.pattern.am[day.get('day') - 1] + employment.pattern.pm[day.get('day') - 1];

export function calculateWorkload(from: string, to: string, activities: MocoActivity[], employments: MocoEmployment[], schedules: MocoSchedule[]): WorkloadType {

    if (!employments.length) {
        return;
    }

    return createDayRange(dayjs(from), dayjs(to))
        .filter(isWeekday)
        .map((day: Dayjs) => {
            const workedHours = activities
                .filter(activity => activity.date === day.format(DATE_FORMAT))
                .reduce((workedHours, activity) => workedHours + activity.hours, 0);

            const employment: MocoEmployment = employments
                .find(({ from, to }: MocoEmployment) =>
                    (!from || day.isSame(from) || day.isAfter(from)) && (!to || day.isSame(to) || day.isBefore(to)));

            const absenceHours = extractAbsenceHours(schedules, null, day);
            // not every time range has an employment; see Lucas Meurer from 28.12.20 - 03.01.21
            const scheduledHours = employment ? extractExpectedHours(employment, day) : 0;

            return {
                day: day.format(DATE_FORMAT),
                expectedHours: Math.max(scheduledHours - absenceHours, 0), // result can be below 0 for part time employees
                worked: workedHours,
                holidays: extractAbsenceHours(schedules, 'Feiertag', day),
                vacations: extractAbsenceHours(schedules, 'Urlaub', day),
                notPlannedHours: extractAbsenceHours(schedules, 'Nicht planbar', day),
                sick: extractAbsenceHours(schedules, 'Krankheit', day),
                absence: extractAbsenceHours(schedules, 'Abwesenheit', day),
            };
        })
        .reduce((workload: WorkloadType, dayObject: DayObjectType & { holidays: number, vacations: number, notPlannedHours: number, sick: number, absence: number }) => {
            const expectedHours = workload.expectedHours + dayObject.expectedHours;
            const workedHours = workload.workedHours + dayObject.worked;
            return {
                ...workload,
                expectedHours: expectedHours,
                workedHours: workedHours,
                percentage: !expectedHours ? 100 : Math.round(workedHours / expectedHours * 100), // expectedHours could be 0 which is not allowed in divisions
                days: [...workload.days, { day: dayObject.day, worked: dayObject.worked, expectedHours: dayObject.expectedHours }],
                absence: workload.absence + dayObject.absence,
                vacations: workload.vacations + dayObject.vacations,
                holidays: workload.holidays + dayObject.holidays,
                notPlanned: workload.notPlanned + dayObject.notPlannedHours,
                sick: workload.sick + dayObject.sick,
            };
        }, {
            user: employments[0].user,
            expectedHours: 0,
            workedHours: 0,
            days: [],
            absence: 0,
            holidays: 0,
            notPlanned: 0,
            sick: 0,
            vacations: 0,
            percentage: 1
        });
}
