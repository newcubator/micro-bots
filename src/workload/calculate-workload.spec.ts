import { calculateWorkload } from "./calculate-workload";

describe("Workload Calculation", () => {
  it("should calculate the workload for one day", () => {
    const from = "2021-01-01"; // friday
    const to = "2021-01-01"; // friday
    const activities: any[] = [{ date: "2021-01-01", hours: 8 }];
    const employments: any[] = [
      {
        from: "2020-01-01",
        to: "2022-01-01",
        pattern: {
          am: [4, 4, 4, 4, 4],
          pm: [4, 4, 4, 4, 4],
        },
        user: { id: "92ee20b5-0721-4d43-8d60-6dc87ccc50e8" /* id is enough for testing*/ },
      },
    ];
    const schedules: any[] = [];

    const workload = calculateWorkload(from, to, activities, employments, schedules);

    expect(workload).toMatchSnapshot();
  });

  it("should calculate the workload for one day and multiple activities", () => {
    const from = "2021-01-01"; // friday
    const to = "2021-01-01"; // friday
    const activities: any[] = [
      { date: "2021-01-01", hours: 1 },
      { date: "2021-01-01", hours: 3 },
      { date: "2021-01-01", hours: 4 },
    ];
    const employments: any[] = [
      {
        from: "2020-01-01",
        to: "2022-01-01",
        pattern: {
          am: [4, 4, 4, 4, 4],
          pm: [4, 4, 4, 4, 4],
        },
        user: { id: "92ee20b5-0721-4d43-8d60-6dc87ccc50e8" /* id is enough for testing*/ },
      },
    ];
    const schedules: any[] = [];

    const workload = calculateWorkload(from, to, activities, employments, schedules);

    expect(workload).toMatchSnapshot();
  });

  it("should calculate the workload for one day for a half time employee", () => {
    const from = "2021-01-01"; // friday
    const to = "2021-01-01"; // friday
    const activities: any[] = [
      { date: "2021-01-01", hours: 1 },
      { date: "2021-01-01", hours: 3 },
    ];
    const employments: any[] = [
      {
        from: "2020-01-01",
        to: "2022-01-01",
        pattern: {
          am: [4, 4, 4, 4, 4],
          pm: [0, 0, 0, 0, 0],
        },
        user: { id: "92ee20b5-0721-4d43-8d60-6dc87ccc50e8" /* id is enough for testing*/ },
      },
    ];
    const schedules: any[] = [];

    const workload = calculateWorkload(from, to, activities, employments, schedules);

    expect(workload).toMatchSnapshot();
  });

  it("should calculate the workload for one day with no activities", () => {
    const from = "2021-01-01"; // friday
    const to = "2021-01-01"; // friday
    const activities: any[] = [];
    const employments: any[] = [
      {
        from: "2020-01-01",
        to: "2022-01-01",
        pattern: {
          am: [4, 4, 4, 4, 4],
          pm: [4, 4, 4, 4, 4],
        },
        user: { id: "92ee20b5-0721-4d43-8d60-6dc87ccc50e8" /* id is enough for testing*/ },
      },
    ];
    const schedules: any[] = [];

    const workload = calculateWorkload(from, to, activities, employments, schedules);

    expect(workload).toMatchSnapshot();
  });

  it("should calculate the workload for one day with absences", () => {
    const from = "2021-01-01"; // friday
    const to = "2021-01-01"; // friday
    const activities: any[] = [];
    const employments: any[] = [
      {
        from: "2020-01-01",
        to: "2022-01-01",
        pattern: {
          am: [4, 4, 4, 4, 4],
          pm: [4, 4, 4, 4, 4],
        },
        user: { id: "92ee20b5-0721-4d43-8d60-6dc87ccc50e8" /* id is enough for testing*/ },
      },
    ];
    const schedules: any[] = [
      {
        date: "2021-01-01",
        assignment: {
          name: "Urlaub",
        },
        am: true,
        pm: true,
      },
    ];

    const workload = calculateWorkload(from, to, activities, employments, schedules);

    expect(workload).toMatchSnapshot();
  });

  it("should calculate the workload for one week", () => {
    const from = "2021-01-01"; // friday
    const to = "2021-01-08"; // friday
    const activities: any[] = [
      { date: "2021-01-01", hours: 8 },
      { date: "2021-01-04", hours: 8 },
      { date: "2021-01-05", hours: 8 },
      { date: "2021-01-06", hours: 8 },
      { date: "2021-01-07", hours: 8 },
      { date: "2021-01-08", hours: 8 },
    ];
    const employments: any[] = [
      {
        from: "2020-01-01",
        to: "2022-01-01",
        pattern: {
          am: [4, 4, 4, 4, 4],
          pm: [4, 4, 4, 4, 4],
        },
        user: { id: "92ee20b5-0721-4d43-8d60-6dc87ccc50e8" /* id is enough for testing*/ },
      },
    ];
    const schedules: any[] = [];

    const workload = calculateWorkload(from, to, activities, employments, schedules);

    expect(workload).toMatchSnapshot();
  });

  it("should calculate the workload for one week with overtime and undertime", () => {
    const from = "2021-01-01"; // friday
    const to = "2021-01-08"; // friday
    const activities: any[] = [
      { date: "2021-01-01", hours: 10 },
      { date: "2021-01-04", hours: 8 },
      { date: "2021-01-05", hours: 8 },
      { date: "2021-01-06", hours: 8 },
      { date: "2021-01-07", hours: 6 },
      { date: "2021-01-08", hours: 8 },
    ];
    const employments: any[] = [
      {
        from: "2020-01-01",
        to: "2022-01-01",
        pattern: {
          am: [4, 4, 4, 4, 4],
          pm: [4, 4, 4, 4, 4],
        },
        user: { id: "92ee20b5-0721-4d43-8d60-6dc87ccc50e8" /* id is enough for testing*/ },
      },
    ];
    const schedules: any[] = [];

    const workload = calculateWorkload(from, to, activities, employments, schedules);

    expect(workload).toMatchSnapshot();
  });

  it("should calculate the workload for one week for part time employees", () => {
    const from = "2021-01-01"; // friday
    const to = "2021-01-08"; // friday
    const activities: any[] = [
      { date: "2021-01-01", hours: 4 },
      { date: "2021-01-04", hours: 4 },
      { date: "2021-01-05", hours: 4 },
      { date: "2021-01-06", hours: 8 },
      { date: "2021-01-07", hours: 0 },
      { date: "2021-01-08", hours: 4 },
    ];
    const employments: any[] = [
      {
        from: "2020-01-01",
        to: "2022-01-01",
        pattern: {
          am: [4, 0, 4, 0, 4],
          pm: [0, 4, 4, 0, 0],
        },
        user: { id: "92ee20b5-0721-4d43-8d60-6dc87ccc50e8" /* id is enough for testing*/ },
      },
    ];
    const schedules: any[] = [];

    const workload = calculateWorkload(from, to, activities, employments, schedules);

    expect(workload).toMatchSnapshot();
  });

  it("should calculate the workload for one week with absences", () => {
    const from = "2021-01-01"; // friday
    const to = "2021-01-08"; // friday
    const activities: any[] = [{ date: "2021-01-01", hours: 8 }];
    const employments: any[] = [
      {
        from: "2020-01-01",
        to: "2022-01-01",
        pattern: {
          am: [4, 4, 4, 4, 4],
          pm: [4, 4, 4, 4, 4],
        },
        user: { id: "92ee20b5-0721-4d43-8d60-6dc87ccc50e8" /* id is enough for testing*/ },
      },
    ];
    const schedules: any[] = [
      { date: "2021-01-04", assignment: { name: "Feiertag" }, am: true, pm: true },
      { date: "2021-01-05", assignment: { name: "Urlaub" }, am: true, pm: true },
      { date: "2021-01-06", assignment: { name: "Nicht planbar" }, am: true, pm: true },
      { date: "2021-01-07", assignment: { name: "Krankheit" }, am: true, pm: true },
      { date: "2021-01-08", assignment: { name: "Abwesenheit" }, am: true, pm: true },
    ];

    const workload = calculateWorkload(from, to, activities, employments, schedules);

    expect(workload).toMatchSnapshot();
  });

  it("should calculate the workload and respect the given time frame", () => {
    const from = "2021-01-01"; // friday
    const to = "2021-01-04"; // monday
    const activities: any[] = [
      { date: "2021-01-01", hours: 8 },
      { date: "2021-01-04", hours: 8 },
      { date: "2021-12-31", hours: 1 }, // not in time frame
      { date: "2021-01-05", hours: 1 }, // not in time frame
    ];
    const employments: any[] = [
      {
        from: "2020-01-01",
        to: "2022-01-01",
        pattern: {
          am: [4, 4, 4, 4, 4],
          pm: [4, 4, 4, 4, 4],
        },
        user: { id: "92ee20b5-0721-4d43-8d60-6dc87ccc50e8" /* id is enough for testing*/ },
      },
    ];
    const schedules: any[] = [
      { date: "2021-01-05", assignment: { name: "Urlaub" }, am: true, pm: true }, // not in time frame
      { date: "2021-12-31", assignment: { name: "Urlaub" }, am: true, pm: true }, // not in time frame
    ];

    const workload = calculateWorkload(from, to, activities, employments, schedules);

    expect(workload).toMatchSnapshot();
  });
});
