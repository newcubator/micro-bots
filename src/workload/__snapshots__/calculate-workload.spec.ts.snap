// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workload Calculation should calculate the workload and respect the given time frame 1`] = `
Object {
  "absence": 0,
  "days": Array [
    Object {
      "day": "2021-01-01",
      "expectedHours": 8,
      "worked": 8,
    },
    Object {
      "day": "2021-01-04",
      "expectedHours": 8,
      "worked": 8,
    },
  ],
  "expectedHours": 16,
  "holidays": 0,
  "notPlanned": 0,
  "percentage": 100,
  "sick": 0,
  "user": Object {
    "id": "92ee20b5-0721-4d43-8d60-6dc87ccc50e8",
  },
  "vacations": 0,
  "workedHours": 16,
}
`;

exports[`Workload Calculation should calculate the workload for one day 1`] = `
Object {
  "absence": 0,
  "days": Array [
    Object {
      "day": "2021-01-01",
      "expectedHours": 8,
      "worked": 8,
    },
  ],
  "expectedHours": 8,
  "holidays": 0,
  "notPlanned": 0,
  "percentage": 100,
  "sick": 0,
  "user": Object {
    "id": "92ee20b5-0721-4d43-8d60-6dc87ccc50e8",
  },
  "vacations": 0,
  "workedHours": 8,
}
`;

exports[`Workload Calculation should calculate the workload for one day and multiple activities 1`] = `
Object {
  "absence": 0,
  "days": Array [
    Object {
      "day": "2021-01-01",
      "expectedHours": 8,
      "worked": 8,
    },
  ],
  "expectedHours": 8,
  "holidays": 0,
  "notPlanned": 0,
  "percentage": 100,
  "sick": 0,
  "user": Object {
    "id": "92ee20b5-0721-4d43-8d60-6dc87ccc50e8",
  },
  "vacations": 0,
  "workedHours": 8,
}
`;

exports[`Workload Calculation should calculate the workload for one day for a half time employee 1`] = `
Object {
  "absence": 0,
  "days": Array [
    Object {
      "day": "2021-01-01",
      "expectedHours": 4,
      "worked": 4,
    },
  ],
  "expectedHours": 4,
  "holidays": 0,
  "notPlanned": 0,
  "percentage": 100,
  "sick": 0,
  "user": Object {
    "id": "92ee20b5-0721-4d43-8d60-6dc87ccc50e8",
  },
  "vacations": 0,
  "workedHours": 4,
}
`;

exports[`Workload Calculation should calculate the workload for one day with absences 1`] = `
Object {
  "absence": 0,
  "days": Array [
    Object {
      "day": "2021-01-01",
      "expectedHours": 0,
      "worked": 0,
    },
  ],
  "expectedHours": 0,
  "holidays": 0,
  "notPlanned": 0,
  "percentage": 100,
  "sick": 0,
  "user": Object {
    "id": "92ee20b5-0721-4d43-8d60-6dc87ccc50e8",
  },
  "vacations": 8,
  "workedHours": 0,
}
`;

exports[`Workload Calculation should calculate the workload for one day with no activities 1`] = `
Object {
  "absence": 0,
  "days": Array [
    Object {
      "day": "2021-01-01",
      "expectedHours": 8,
      "worked": 0,
    },
  ],
  "expectedHours": 8,
  "holidays": 0,
  "notPlanned": 0,
  "percentage": 0,
  "sick": 0,
  "user": Object {
    "id": "92ee20b5-0721-4d43-8d60-6dc87ccc50e8",
  },
  "vacations": 0,
  "workedHours": 0,
}
`;

exports[`Workload Calculation should calculate the workload for one week 1`] = `
Object {
  "absence": 0,
  "days": Array [
    Object {
      "day": "2021-01-01",
      "expectedHours": 8,
      "worked": 8,
    },
    Object {
      "day": "2021-01-04",
      "expectedHours": 8,
      "worked": 8,
    },
    Object {
      "day": "2021-01-05",
      "expectedHours": 8,
      "worked": 8,
    },
    Object {
      "day": "2021-01-06",
      "expectedHours": 8,
      "worked": 8,
    },
    Object {
      "day": "2021-01-07",
      "expectedHours": 8,
      "worked": 8,
    },
    Object {
      "day": "2021-01-08",
      "expectedHours": 8,
      "worked": 8,
    },
  ],
  "expectedHours": 48,
  "holidays": 0,
  "notPlanned": 0,
  "percentage": 100,
  "sick": 0,
  "user": Object {
    "id": "92ee20b5-0721-4d43-8d60-6dc87ccc50e8",
  },
  "vacations": 0,
  "workedHours": 48,
}
`;

exports[`Workload Calculation should calculate the workload for one week for part time employees 1`] = `
Object {
  "absence": 0,
  "days": Array [
    Object {
      "day": "2021-01-01",
      "expectedHours": 4,
      "worked": 4,
    },
    Object {
      "day": "2021-01-04",
      "expectedHours": 4,
      "worked": 4,
    },
    Object {
      "day": "2021-01-05",
      "expectedHours": 4,
      "worked": 4,
    },
    Object {
      "day": "2021-01-06",
      "expectedHours": 8,
      "worked": 8,
    },
    Object {
      "day": "2021-01-07",
      "expectedHours": 0,
      "worked": 0,
    },
    Object {
      "day": "2021-01-08",
      "expectedHours": 4,
      "worked": 4,
    },
  ],
  "expectedHours": 24,
  "holidays": 0,
  "notPlanned": 0,
  "percentage": 100,
  "sick": 0,
  "user": Object {
    "id": "92ee20b5-0721-4d43-8d60-6dc87ccc50e8",
  },
  "vacations": 0,
  "workedHours": 24,
}
`;

exports[`Workload Calculation should calculate the workload for one week with absences 1`] = `
Object {
  "absence": 8,
  "days": Array [
    Object {
      "day": "2021-01-01",
      "expectedHours": 8,
      "worked": 8,
    },
    Object {
      "day": "2021-01-04",
      "expectedHours": 0,
      "worked": 0,
    },
    Object {
      "day": "2021-01-05",
      "expectedHours": 0,
      "worked": 0,
    },
    Object {
      "day": "2021-01-06",
      "expectedHours": 0,
      "worked": 0,
    },
    Object {
      "day": "2021-01-07",
      "expectedHours": 0,
      "worked": 0,
    },
    Object {
      "day": "2021-01-08",
      "expectedHours": 0,
      "worked": 0,
    },
  ],
  "expectedHours": 8,
  "holidays": 8,
  "notPlanned": 8,
  "percentage": 100,
  "sick": 8,
  "user": Object {
    "id": "92ee20b5-0721-4d43-8d60-6dc87ccc50e8",
  },
  "vacations": 8,
  "workedHours": 8,
}
`;

exports[`Workload Calculation should calculate the workload for one week with overtime and undertime 1`] = `
Object {
  "absence": 0,
  "days": Array [
    Object {
      "day": "2021-01-01",
      "expectedHours": 8,
      "worked": 10,
    },
    Object {
      "day": "2021-01-04",
      "expectedHours": 8,
      "worked": 8,
    },
    Object {
      "day": "2021-01-05",
      "expectedHours": 8,
      "worked": 8,
    },
    Object {
      "day": "2021-01-06",
      "expectedHours": 8,
      "worked": 8,
    },
    Object {
      "day": "2021-01-07",
      "expectedHours": 8,
      "worked": 6,
    },
    Object {
      "day": "2021-01-08",
      "expectedHours": 8,
      "worked": 8,
    },
  ],
  "expectedHours": 48,
  "holidays": 0,
  "notPlanned": 0,
  "percentage": 100,
  "sick": 0,
  "user": Object {
    "id": "92ee20b5-0721-4d43-8d60-6dc87ccc50e8",
  },
  "vacations": 0,
  "workedHours": 48,
}
`;
